### DISCUSSION TITLE
Drawing DiGraph & MultiDiGraph multiple edges and labels

### DISCUSSION THREAD

**Discussion by dg-pb** [networkx/networkx#7008] [Q&A]
Hi everyone,

I am working with drawing multiple edges and labels of MultiDiGraph.

DiGraph can draw 2 edges, but can not draw labels for them. To do that I have found replacement for `nx.draw_networkx_edge_labels`
in stack: https://stackoverflow.com/a/70245742/1933583

It was easy to extend it to MultiDiGraph: with few lines, it can take a sequence of radians and labels can be drawn for MultiDiGraph.

So what remains is Drawing edges from MultiDiGraph.

Is there a way to do these already and I just couldn't find it?

Or if there is not way for that, then would suggestion as follows be sensible?

Functions:
a) draw_networkx_edges
b) draw_networkx_edge_labels

Can take a sequence for argument `rad` and correctly draw arbitrary amount of edges and labels.

By the way, my current graph fro MultiDiGraph looks like that:
[MultiDiGraph](https://pasteboard.co/NDmJ6PGEUa8u.png)

As you can see, labels are correctly drawn, but I still haven't written edge drawing.

**Comment by dschult** [comment]
Good places to look for previous work and attempts to work with multiedge drawing in NetworkX are:
#3813 
#5882 
#6806 
I think we have made more progress with multi-edge drawing than with multi-edge labels. So you are in a good position to put some of these improvements together.

**Comment by dg-pb** [comment]
So I have managed to make it work with fairly few lines of code. Looks like this now:

![](https://gcdnb.pbrd.co/images/Odf2DBEwiaju.png?o=1)

Label function change looks as follows:
```python
def my_draw_networkx_edge_labels(
    G,
    pos,
    edge_labels=None,
   ...
   rad=0.1,
):
...
    if edge_labels is None:
        labels = {(u, v): d for u, v, d in G.edges(data=True)}
    else:
        labels = edge_labels
    text_items = {}
    rad = IterMem(itt.accumulate(iterate(rad, n=-1)))
    # this is simply an iterator with memory and allows list indexing
    # which wraps accumulation of infinite scalar
    for nodes, label in labels.items():
        if isinstance(G, nx.MultiDiGraph):
            n1, n2, i = nodes
            rad_ = rad[i]
        else:
            n1, n2 = nodes
            rad_ = rad
        (x1, y1) = pos[n1]
        (x2, y2) = pos[n2]
   ...
```

I plotted the edges themselves by:

```python
def draw_edges(G, pos, ax, rad=0, **kwds):
    if isinstance(G, nx.MultiDiGraph):
        edges = list(G.edges(keys=True, data=False))
        d = defaultdict(list)
        for *e, eid in edges:
            d[eid].append(e)
        rad = itt.accumulate(iterate(rad, n=len(d)))
        for (eid, edge_group), r in zip(d.items(), rad):
            if r != 0:
                kwds['connectionstyle'] = f'arc3, rad={r}'
            nx.draw_networkx_edges(G, pos, edgelist=edge_group, ax=ax, **kwds)
    else:
        edges = list(G.edges(data=False))
        if rad != 0:
            kwds['connectionstyle'] = f'arc3, rad={rad_}'
        nx.draw_networkx_edges(G, pos, edgelist=partial_edges, ax=ax, **kwds)
```

But I believe it could be brought inside the function in the same way as for labels. If this approach looks sensible to you I could issue PR.

One question that I am sure I can not make myself is about rad input. It is either:
a) Strictly Number | list | tuple, where i'th item is `rad` for i`th edge_id
b) Input is a scalar, which is accumulated as many times as needed

For edge lines, it is a bit trickier as it is part of `connectionstyle` argument. Here it is either:
a) `connectionstyle` is strictly str | list | tuple, where i'th item is applied to i'th edge_id
b) extra rad argument is introduced and it is appended to connection style. Here the above question regarding numeric input remains as in labels.

I think a) and a) for both is safer.

On a side note.
Netgraph library does a nice work in plotting MultiEdges. It's working in dev version now:
![](https://gcdnb.pbrd.co/images/tMfMz1HnsooC.png?o=1)

But I still want to have a good fallback in networkx itself for cases where minimal dependencies are preferred. And for the cases where computation time is important - netgraph is pretty hefty - 0.028 vs 0.187 for a 9-edged graph.

**Comment by dg-pb** [comment]
See PR https://github.com/networkx/networkx/pull/7010

**Comment by dg-pb** [comment]
I will just copy-paste the e-mail I have sent to mailing group as it seems it isn't very active.

--------
Hi all,

I have done a bit of work and issued PR: https://github.com/networkx/networkx/pull/7010

It is not breaking any backwards compatibility and is ready to be merged. Could someone take a look and let me know if there are any issues with it or reasons why it can't (will not) be merged?

Secondly,

I was about to go down the path that I have laid out in the comment of making edge drawing and edge label drawing in sync. And incorporate work that has been done in https://github.com/networkx/networkx/issues/3813.

To achieve this sync, some refactoring is needed and certain functionality needs to be factored out from `currently independent on anything else` drawing functions. Also, some changes are inevitable. I think it is possible to avoid breaking backwards compatibility, but changes, such as self-loop height decoupling from certain parameters, are inevitable.

So I am a bit reluctant to go forward with this as I am not sure what boundaries I am in and I would need some guidance in this respect.

Also, I personally do not have any need for it as what I have done already is sufficient for my own needs. So I would only take this on if I was sure that my work is going to be reviewed and someone will make sure it is merged in the end.

Regards,
DG

**Reply by dschult** [reply]
I like what I see here. This has been a problem for a while and (as you point out) solutions might require rethinking parts of the drawing API.  I'll put more comments on the PR itself, but in terms of answering your "Secondly" question: 
- I think it would be good to have the first PR improve the current situation with minimal api changes. Of course other PRs in this category are possible too. It doesn't have to all be in one PR.
- Maybe make a separate PR suggesting changes which break the current api but end up with a better api.  

We are open to the idea of making breaking changes that lead to a new better system. Refactoring is definitely ok with factored out helper functions (names starting with `_` please, unless you expect them to be public functions) called by both the edges and the labels code. 
Are you talking about combining drawing edges with drawing labels? Or just factoring out common parts? 

If you need info about nodes or about edges when doing other parts, what would be a good way to store that info for either the other drawing function, or for later additions of other nodes/edges? I guess the node drawing function could return an object that could be passed to other functions. I've heard of matplotlib Artists and perhaps framing this around a networkx focused artist object could be one way to keep all the info together.  Lots of possibilities.  Start simple... Start somewhere.  What I've seen here and in the PR looks good.

**Reply by dg-pb** [reply]
I already took the liberty and did the work without breaking any backwards compatibility. It is ready for review. See my last comment in PR.
